Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2017 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
+#include "CryptoAlgorithmEd25519.h"
+
+#if ENABLE(WEB_CRYPTO)
+#include "CommonCryptoDERUtilities.h"
+#include "CommonCryptoUtilities.h"
+#include "CryptoAlgorithmEd25519Params.h"
+#include "CryptoDigestAlgorithm.h"
+#include "CryptoKeyEC.h"
+#include <pal/spi/cocoa/CoreCryptoSPI.h>
+#include <stdio.h>
+#include <string.h>
+#include <wtf/Assertions.h>
+#include <wtf/text/Base64.h>
+
+namespace WebCore {
+static ExceptionOr<Vector<uint8_t>> signEd25519(CryptoAlgorithmIdentifier, const PlatformECKey key, size_t , const Vector<uint8_t>& data)
+{
+    Vector<uint8_t> sk = std::get<Vector<uint8_t>>(key);
+    ccec25519pubkey pk;
+    const struct ccdigest_info* di = ccsha512_di();
+    cced25519_make_pub(di, pk, sk.data());
+    Vector<uint8_t> publicKey(pk, 32);
+    ccec25519signature newSignature = { };
+    cced25519_sign(di, newSignature, data.size(), data.data(), pk, sk.data());
+    Vector<uint8_t> sg(newSignature, 64);
+    return WTFMove( sg);
+
+}
+
+static ExceptionOr<bool> verifyEd25519(CryptoAlgorithmIdentifier , const PlatformECKey key, size_t keyLengthInBytes, const Vector<uint8_t>& signature, const Vector<uint8_t> data)
+{
+    Vector<uint8_t> pk = std::get<Vector<uint8_t>>(key);
+    if (signature.size() != keyLengthInBytes * 2)
+        return false;
+    const struct ccdigest_info* di = ccsha512_di();
+    int result = cced25519_verify(di, data.size(), data.data(), signature.data(), pk.data());
+    if (result)
+        return false;
+    return true;
```

Comment:
You can use
return !cced25519_verify(di, data.size(), data.data(), signature.data(), pk.data())